cpp/ql/src/experimental/Security/CWE/CWE-783/OperatorPrecedenceLogicErrorWhenUseBitwiseOrLogicalOperations.ql

```diff
@@ -0,0 +1,179 @@
+    not exptmp
+        .(LogicalOrExpr)
+        .getRightOperand()
+        .(BinaryBitwiseOperation)
+        .getRightOperand()
+        .getValue() = "0" and
+    not exptmp
+        .(LogicalOrExpr)
+        .getRightOperand()
+        .(BinaryBitwiseOperation)
+        .getRightOperand()
+        .getValue() = "1"
+  )
+  or
+  not exptmp.(LogicalAndExpr).getLeftOperand() instanceof BinaryOperation and
+  not exptmp.(LogicalAndExpr).getRightOperand().isParenthesised() and
+  (
+    exptmp.(LogicalAndExpr).getRightOperand().(BinaryBitwiseOperation).getLeftOperand().getType()
+      instanceof BoolType and
+    not exptmp
+        .(LogicalAndExpr)
+        .getRightOperand()
+        .(BinaryBitwiseOperation)
+        .getRightOperand()
+        .getValue() = "0" and
+    not exptmp
+        .(LogicalAndExpr)
+        .getRightOperand()
+        .(BinaryBitwiseOperation)
+        .getRightOperand()
+        .getValue() = "1"
+  )
+}
+
+/** Holds if `exptmp` equals expression bit operations in reverse priority order. */
+predicate isBitwiseandBitwise(Expr exptmp) {
+  not exptmp.(BitwiseOrExpr).getLeftOperand() instanceof BinaryOperation and
+  not exptmp.(BitwiseOrExpr).getRightOperand().isParenthesised() and
+  (
+    exptmp.(BitwiseOrExpr).getRightOperand() instanceof BitwiseAndExpr or
+    exptmp.(BitwiseOrExpr).getRightOperand() instanceof BitwiseXorExpr
+  )
+  or
+  not exptmp.(BitwiseXorExpr).getLeftOperand() instanceof BinaryOperation and
+  not exptmp.(BitwiseXorExpr).getRightOperand().isParenthesised() and
+  exptmp.(BitwiseXorExpr).getRightOperand() instanceof BitwiseAndExpr
+}
+
+/** Holds if the range contains no boundary values. */
+predicate isRealRange(Expr exp) {
+  upperBound(exp).toString() != "18446744073709551616" and
+  upperBound(exp).toString() != "9223372036854775807" and
+  upperBound(exp).toString() != "4294967295" and
+  upperBound(exp).toString() != "Infinity" and
+  upperBound(exp).toString() != "NaN" and
+  lowerBound(exp).toString() != "-9223372036854775808" and
+  lowerBound(exp).toString() != "-4294967296" and
+  lowerBound(exp).toString() != "-Infinity" and
+  lowerBound(exp).toString() != "NaN" and
+  upperBound(exp) != 2147483647 and
+  upperBound(exp) != 268435455 and
+  upperBound(exp) != 33554431 and
+  upperBound(exp) != 8388607 and
+  upperBound(exp) != 65535 and
+  upperBound(exp) != 32767 and
+  upperBound(exp) != 255 and
+  upperBound(exp) != 127 and
+  lowerBound(exp) != -2147483648 and
+  lowerBound(exp) != -268435456 and
+  lowerBound(exp) != -33554432 and
+  lowerBound(exp) != -8388608 and
+  lowerBound(exp) != -65536 and
+  lowerBound(exp) != -32768 and
+  lowerBound(exp) != -128
+  or
+  lowerBound(exp) = 0 and
+  upperBound(exp) = 1
+}
+
+/** Holds if expressions are of different size or range */
+predicate isDifferentSize(Expr exp1, Expr exp2, Expr exp3) {
+  exp1.getType().getSize() = exp2.getType().getSize() and
+  exp1.getType().getSize() != exp3.getType().getSize()
```

Comment:
The first two lines of this predicate says: `isDifferentSize` includes all tuples `(exp1, exp2, exp3)` where the size of the type of `exp1` is equal to the size of the type of `exp2`, and the size of the type of `exp1` is different from the size of the type of `exp3`. That is _a lot_ of tuples! In fact, I couldn't actually evaluate this predicate on any real projects because the result of this predicate is too large.

There are many ways to fix this. The easiest way to fix it is to tell QL that you don't want this predicate evaluated in isolation. Instead, you want to inline the predicate at all of the call sites. You can do that with the line `pragma[inline]`. Then the giant result is never created.

If you've noticed that your code was too slow to run on some project, try to include this `pragma[inline]` line and run it again. I'm pretty sure you'll notice a big performance boost :)

```suggestion
pragma[inline]
predicate isDifferentSize(Expr exp1, Expr exp2, Expr exp3) {
  exp1.getType().getSize() = exp2.getType().getSize() and
  exp1.getType().getSize() != exp3.getType().getSize()
```

Reply:
thanks for this idea.
I will definitely use it.
I'm just going to run some tests.
Thank you.